src/query/storage/m3/cluster_resolver.go

```diff
@@ -0,0 +1,299 @@
+	// First check if the unaggregated cluster can fully satisfy the query range.
+	// If so, return it and shortcircuit, as unaggregated will necessarily have
+	// every metric.
+	unaggregated := resolveUnaggregatedNamespaceForQuery(now, start,
+		clusters.UnaggregatedClusterNamespace(), opts)
+	if unaggregated.satisfies == fullySatisfiesRange {
+		return namespaceCoversAllQueryRange,
+			ClusterNamespaces{unaggregated.clusterNamespace},
+			nil
+	}
+
+	// The filter function will drop namespaces which do not cover the entire
+	// query range from contention.
+	//
+	// NB: if fanout aggregation is forced on, the filter instead forces clusters
+	// that do not cover the range to be set as partially aggregated.
+	filterFunc := func(namespace ClusterNamespace) bool {
+		// Include only if can fulfill the entire time range of the query
+		clusterStart := now.Add(-1 * namespace.Options().Attributes().Retention)
+		return !clusterStart.After(start)
+	}
+
+	// Filter aggregated namespaces by filter function and options.
+	var r reusedAggregatedNamespaceSlices
+	r = aggregatedNamespaces(clusters.ClusterNamespaces(), r, filterFunc, opts)
+
+	// If any of the aggregated clusters have a complete set of metrics, use
+	// those that have the smallest resolutions, supplemented by lower resolution
+	// partially aggregated metrics.
+	if len(r.completeAggregated) > 0 {
+		sort.Stable(ClusterNamespacesByResolutionAsc(r.completeAggregated))
+		// Take most granular complete aggregated namespace.
+		result := r.completeAggregated[:1]
+		completedAttrs := result[0].Options().Attributes()
+		// Also include any finer grain partially aggregated namespaces that
+		// may contain a matching metric.
+		for _, n := range r.partialAggregated {
+			if n.Options().Attributes().Resolution < completedAttrs.Resolution {
+				// More granular resolution.
+				result = append(result, n)
+			}
+		}
+
+		return namespaceCoversAllQueryRange, result, nil
+	}
+
+	// No need to regenerate aggregated namespace without a filter if aggregated
+	// fanout is force enabled.
+	if opts.FanoutAggregated != storage.FanoutForceEnable {
+		// No complete aggregated namespaces can definitely fulfill the query,
+		// so take the longest retention completed aggregated namespace to return
+		// as much data as possible, along with any partially aggregated namespaces
+		// that have either same retention and lower resolution or longer retention
+		// than the complete aggregated namespace
+		r = aggregatedNamespaces(clusters.ClusterNamespaces(), r, nil, opts)
+	}
+
+	if len(r.completeAggregated) == 0 {
+		// Absolutely no complete aggregated namespaces, need to fanout to all
+		// partial aggregated namespaces as well as the unaggregated cluster
+		// as we have no idea which has the longest retention.
+		result := r.partialAggregated
+		// If unaggregated namespace can partially satisfy this range, add it as a
+		// fanout contender.
+		if unaggregated.satisfies == partiallySatisfiesRange {
```

Comment:
Is this check necessary?  If it fully satisifies the query range it would've already be returned on line 87 as part of the first check no?

Reply:
Yeah, it's to check that it's not `disabled` here... I didn't want to use two bools like `enabled, coversRange bool` since they're kinda linked